### PR TITLE
[ROCm][AITER] Directly Implement AITER Custom All-reduce in CudaCommunicator

### PR DESCRIPTION
**Pull request by BadrBasowid**
## Purpose
Previously, with the AITER fused AR-RMSNorm enabled, two separate custom-allreduce instances were live: vLLM's `CustomAllreduce` for the plain allreduce, and a standalone AITER `CustomAllreduce` (`rocm_aiter_ops._CUSTOM_ALL_REDUCE`) for the fused op, each with its own IPC buffers, both registered during CUDA graph capture. Performance was the main reason it's done this way rather than by dropping the AITER path directly in the `CudaCommunicator`, because AITER's unfused custom-allreduce kernel was slower than vLLM's.

That method introduced in #37646, however, was flaky and complicated. Since the performance disadvantage that previously justified retaining vLLM's kernel no longer holds (AITER's custom all-reduce has since closed that gap and no longer exhibits the unfused-kernel regression), consolidating onto AITER is now both safe and preferable.

The major changes are summarized below:
- New env flag ~`VLLM_ROCM_USE_AITER_CUSTOM_ALL_REDUCE`~ `VLLM_ROCM_USE_AITER_CUSTOM_AR` to enable Aiter's custom allreduce.
- Replace `AiterCustomAllreduceProto` protocol with a simple `AiterCustomAllreduce` class that wraps Aiter's CA impl.
- CudaCommunicator inits an `aiter_ar_comm` backend (in a similar way as FlashInfer inits its CA backend).

Perf results are listed below:

kimi

dsv3

ds-v4

llama-tp2

llama-tp4

qwen-tp2

qwen-tp4

Averaging these results based on TP:

Model | TP | ITL | TTFT | TPOT
-- | -- | -- | -- | --
Llama 3.1 8B | tp2 | +0.60% | -2.10% | +0.60%
Llama 3.1 8B | tp4 | -1.32% | -1.00% | -1.32%
Llama 3.1 8B | overall | -0.36% | -1.55% | -0.36%
Qwen3 30B A3B | tp2 | +0.09% | -0.45% | +0.09%
Qwen3 30B A3B | tp4 | -1.60% | +2.92% | -1.60%
Qwen3 30B A3B | overall | -0.76% | +1.23% | -0.76%

### Additional note
I have also noticed an accuracy regression for `moonshotai/Kimi-K2.5`( down to 0.4 on lm_eval gsm8k) on upstream when using `flydsl` MOE backend with `fuse_allreduce_rms` turned ON.  Setting `AMD_SERIALIZE_COPY=3` fixed it, which indicates upstream `fuse_allreduce_rms` likely had a synchronization issue. This PR also fixes this issues and lm_eval score recovers to normal level with the `flydsl` MOE backend.

## Test Plan

1. `pytest vllm/tests/distributed/test_rocm_aiter_custom_all_reduce.py` (New)
2. `pytest tests/compile/passes/distributed/test_fusion_all_reduce.py`
3. `pytest tests/compile/fusions_e2e/test_tp2_ar_rms.py`
4. gsm8k `moonshotai/Kimi-K2.5` lm_eval with ar_rms fusion
5. gsm8k `moonshotai/Kimi-K2.5` lm_eval without ar_rms fusion
6. CI

## Test Result

1. `== 2 passed, 17 warnings in 34.32s ==`
2. `== 36 passed, 180 skipped, 16 warnings in 1514.87s (0:25:14) ==`
3. `== 4 passed, 20 deselected, 24 warnings in 36.90s ==`
4. flexible-extract = 0.9363 | strict-match = 0.9371
5. flexible-extract = 0.9393 | strict-match = 0.9401

## AI assistance

This PR was developed with AI assistance. All lines were reviewed by me, BadrBasowid, the submitter.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/distributed/device_communicators/aiter_custom_all_reduce.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    def supports_dynamic_hidden_dim(self) -> bool:
+        """Aiter's fused_allreduce_rmsnorm kernel dispatches on hidden_dim.
+        Before aiter v0.1.12 the launcher was template-specialized on HIDDEN_DIM
+        and silently no-op'd for sizes outside {512, 1024, 2048, 4096}. From v0.1.12
+        hidden_dim is a runtime argument. Older builds are detected via
+        AiterCustomAllreduce.supports_dynamic_hidden_dim; This function is used to
+        skip fusion for unsupported sizes on them.
+        Ref (old kernel): https://github.com/ROCm/aiter/blob/6a0e7b26ccf33164785531212cc2ec2cde0b9243/csrc/include/custom_all_reduce.cuh#L2590
+        """
+        return hasattr(self._impl, "_pool")
```
Comment (BadrBasowid) [inline vllm/distributed/device_communicators/aiter_custom_all_reduce.py:75]: This is ugly and flaky(aiter could simply remove `_pool` from the next version), see if there's a better fix

File: vllm/envs.py [hunk 2]
```diff
@@ -1128,6 +1129,12 @@ def _resolve_rust_frontend_path() -> str | None:
     "VLLM_ROCM_USE_AITER": lambda: (
         os.getenv("VLLM_ROCM_USE_AITER", "False").lower() in ("true", "1")
     ),
+    # Use AITER's CustomAllreduce as the custom-allreduce backend inside vLLM's
+    # CudaCommunicator on ROCm.
+    "VLLM_ROCM_USE_AITER_CUSTOM_ALL_REDUCE": lambda: (
```
Comment (BadrBasowid) [inline vllm/envs.py]: @tjtanaa Shall we keep this `True` by default?

File: vllm/distributed/device_communicators/aiter_custom_all_reduce.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+import torch
+from torch.distributed import ProcessGroup
+
+from vllm.logger import init_logger
+
+logger = init_logger(__name__)
+
+
+class AiterCustomAllreduce:
```
Comment (BadrBasowid) [inline vllm/distributed/device_communicators/aiter_custom_all_reduce.py:19]: Add some tests and compare with vLLM's CA

File: tests/compile/fusions_e2e/conftest.py [hunk 4]
```diff
@@ -79,6 +79,9 @@ def run(
     ):
         monkeypatch.setenv("VLLM_USE_DEEP_GEMM", "1" if use_deepgemm else "0")
         monkeypatch.setenv("VLLM_ROCM_USE_AITER", "1" if use_aiter else "0")
+        monkeypatch.setenv(
+            "VLLM_ROCM_USE_AITER_CUSTOM_ALL_REDUCE", "1" if use_aiter else "0"
```
Comment (tjtanaa) [inline tests/compile/fusions_e2e/conftest.py]: Let's use `VLLM_ROCM_USE_AITER_CUSTOM_AR` shorthand

File: vllm/distributed/device_communicators/aiter_custom_all_reduce.py (hunk 1 above)
Comment (BadrBasowid) [reply vllm/distributed/device_communicators/aiter_custom_all_reduce.py:75]: Aiter doesn't expose `__version__` attribute, and using `importlib.metadata.version` might be even flakier. I think it's just better to keep it this way until we pin a higher version.

File: vllm/envs.py (hunk 2 above)
Comment (tjtanaa) [reply vllm/envs.py]: Let's get more data for a few more models.

File: tests/compile/fusions_e2e/conftest.py (hunk 4 above)
Comment (BadrBasowid) [reply tests/compile/fusions_e2e/conftest.py]: Done!

**Comment by BadrBasowid** [comment]
@tjtanaa @AndreasKaratzas This PR is ready for review

**Comment by tjtanaa** [comment]
@BadrBasowid can you help to collect some benchmark for small TPs on smaller models? The data looks good for large TPs this time of upstream effort.

**Comment by tjtanaa** [comment]
Extra information for other reviewers:

AITER custom all reduce has been porting its implementation from vLLM and optimized since then it already has vLLM custom all reduce, Quick reduce etc). So we don't need to add custom dispatching logic where we dispatch between AITER's Custom AllReduce and vLLM's Custom All Reduce. Moreover, the custom all reduce buffers are managed internally in the C++ layer which makes sharing the ops between AITER's CustomAllReduce and vLLM's CustomAllReduce impossible. This has causes bugs like #37646 .

All of these reasons show that we should only just activate AITER CustomAllReduce instead of mixing them.

AITER Custom All Reduce is heavily tested, validated and optimized along side with ATOM. So, its perf and stability for popular models are more polished. The flag is still introduced because there are possibility that the AITER Custom AllReduce is not that optimized from smaller models or smaller TPs, based on past upstreaming efforts (this is not the first PR to upstream this).

File: vllm/distributed/device_communicators/aiter_custom_all_reduce.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        this so vLLM degrades to the AR+RMS-only fusion when run against an
+        older aiter that lacks the per-group launcher.
+        """
+        from aiter.dist.device_communicators.custom_all_reduce import (
+            CustomAllreduce as _AiterCustomAllreduce,
+        )
+
+        return hasattr(_AiterCustomAllreduce, "fused_ar_rms_per_group_quant")
+
+    # The function is equivalent to `has_fused_allreduce_rmsnorm_quant_per_group`
```
Comment (tjtanaa) [inline vllm/distributed/device_communicators/aiter_custom_all_reduce.py]: please clean up this comment.

**Review by tjtanaa — APPROVED** [review]
LGTM. Thanks for the extensive benchmarking to provide a broader view of the state of AITER's Custom All Reduce vs vLLM's Custom All Reduce

**Comment by tjtanaa** [comment]
@BadrBasowid please evaluate the custom all reduce API again now that we have upgrade to aiter v0.1.16.post2

File: tests/distributed/test_rocm_aiter_custom_all_reduce.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+@pytest.mark.parametrize("test_target", [eager_allreduce, graph_allreduce])
+def test_rocm_aiter_custom_allreduce(
+    monkeypatch: pytest.MonkeyPatch,
+    tp_size,
+    pipeline_parallel_size,
+    test_target,
+):
+    world_size = tp_size * pipeline_parallel_size
+    if world_size > torch.accelerator.device_count():
+        pytest.skip("Not enough GPUs to run the test.")
```
Comment (AndreasKaratzas) [inline tests/distributed/test_rocm_aiter_custom_all_reduce.py]: wait is that ever the case? is this ever skipped on CI as the test group is set right now?

File: tests/distributed/test_rocm_aiter_custom_ar.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+from vllm.distributed.parallel_state import get_tp_group, graph_capture
+from vllm.envs import disable_envs_cache
+from vllm.platforms import current_platform
+
+from ..utils import (
+    ensure_model_parallel_initialized,
+    init_test_distributed_environment,
+    multi_process_parallel,
+)
+
```
Comment (AndreasKaratzas) [inline tests/distributed/test_rocm_aiter_custom_ar.py:22]: I am not able to find any small dummy e2e test for this

Comment (tjtanaa) [reply tests/distributed/test_rocm_aiter_custom_ar.py:22]: @AndreasKaratzas what do you think about us enabling this by default. Because from our perf tests, we found that it is overall faster for popular models. And with enabling by default, this feature will also be tested by e2e unit tests that set `VLLM_ROCM_USE_AITER=1` . 

And for any future `fused op + all reduce` will only go through AITER custom all reduce backend. Not vLLM Custom All reduce backend. We will have more perf gain from using fused op

File: tests/distributed/test_rocm_aiter_custom_all_reduce.py (hunk 6 above)
Comment (BadrBasowid) [reply tests/distributed/test_rocm_aiter_custom_all_reduce.py]: Num gpus is set to 2:
```yaml
agent_pool: mi355_2
num_gpus: 2
```
So i do not think CI will encounter such a case, but it still gaurds in cases where anyone runs locally with `tp < 2`. Would you like me to remove this gaurd?

Comment (BadrBasowid) [reply tests/distributed/test_rocm_aiter_custom_all_reduce.py]: I used the `multi_gpu_test` marker instead

File: tests/distributed/test_rocm_aiter_custom_ar.py (hunk 7 above)
Comment (BadrBasowid) [reply tests/distributed/test_rocm_aiter_custom_ar.py:22]: @AndreasKaratzas @tjtanaa I made aiter's CA on by default (still depends on `VLLM_ROCM_USE_AITER` parent switch) and  added an e2e test comparing aiter's ca with vllm's ca

**Review by AndreasKaratzas — APPROVED** [review]
Tests look good.

**Comment by BadrBasowid** [comment]
More lm_eval results below (both TP4):
```
local-completions ({'model': 'meta-llama/Llama-3.1-8B', 'base_url': 'http://0.0.0.0:9050/v1/completions', 'num_concurrent': 256, 'max_retries': 10, 'max_gen_toks': 2048, 'max_length': 1048576, 'timeout': 60000}), gen_kwargs: ({}), limit: None, num_fewshot: 20, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|    20|exact_match|↑  |0.5118|±  |0.0138|
|     |       |strict-match    |    20|exact_match|↑  |0.5125|±  |0.0138|
```

```
local-completions ({'model': 'Qwen/Qwen3-30B-A3B', 'base_url': 'http://0.0.0.0:9050/v1/completions', 'num_concurrent': 256, 'max_retries': 10, 'max_gen_toks': 2048, 'max_length': 1048576, 'timeout': 60000}), gen_kwargs: ({}), limit: None, num_fewshot: 20, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|    20|exact_match|↑  |0.9242|±  |0.0073|
|     |       |strict-match    |    20|exact_match|↑  |0.9242|±  |0.0073|

```

**Comment by BadrBasowid** [comment]
@tjtanaa @AndreasKaratzas both newly added tests pass:
```
================== 4 passed, 16 warnings in 404.87s (0:06:44) ==================

================== 2 passed, 16 warnings in 90.17s (0:01:30) ===================
```
Quick reduce test failures are unrelated.

**Review by tjtanaa — APPROVED** [review]
LGTM. Thanks for adding the test and more evaluation results
